File: lib/cors.ts
/**
 * CORS (Cross-Origin Resource Sharing) Utility
 * 
 * Provides centralized CORS configuration for API routes.
 * Supports configurable allowed origins for production deployments.
 * 
 * Environment Variables:
 * - ALLOWED_ORIGINS: Comma-separated list of allowed origins (e.g., "https://domain1.vercel.app,https://domain2.com")
 * - NEXT_PUBLIC_BASE_URL: Base URL for the application (used as fallback allowed origin)
 * 
 * @module cors
 */

import type { NextApiResponse } from 'next';

/**
 * Get allowed origins from environment variables
 * Falls back to wildcard if not configured
 * 
 * @returns Array of allowed origins or '*' for all origins
 */
export function getAllowedOrigins(): string | string[] {
  // Check for explicit ALLOWED_ORIGINS configuration
  const allowedOriginsEnv = process.env.ALLOWED_ORIGINS;
  if (allowedOriginsEnv) {
    return allowedOriginsEnv.split(',').map(origin => origin.trim());
  }

  // Check for NEXT_PUBLIC_BASE_URL as fallback
  const baseUrl = process.env.NEXT_PUBLIC_BASE_URL;
  if (baseUrl) {
    return [baseUrl];
  }

  // Default to wildcard for development/demo
  return '*';
}

/**
 * Set CORS headers on an API response
 * 
 * @param res - Next.js API response object
 * @param origin - Optional specific origin to allow (overrides config)
 */
export function setCorsHeaders(res: NextApiResponse, origin?: string): void {
  const allowedOrigins = origin || getAllowedOrigins();

  if (allowedOrigins === '*') {
    res.setHeader('Access-Control-Allow-Origin', '*');
  } else if (typeof allowedOrigins === 'string') {
    res.setHeader('Access-Control-Allow-Origin', allowedOrigins);
  } else if (Array.isArray(allowedOrigins) && allowedOrigins.length > 0) {
    // For multiple origins, we'd need to check the request origin
    // For now, use the first one or wildcard
    res.setHeader('Access-Control-Allow-Origin', allowedOrigins[0]);
    res.setHeader('Vary', 'Origin');
  }

  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  res.setHeader('Access-Control-Max-Age', '86400'); // 24 hours
}

/**
 * Handle CORS preflight (OPTIONS) requests
 * 
 * @param res - Next.js API response object
 * @returns true if this was an OPTIONS request, false otherwise
 */
export function handleCorsPrelight(res: NextApiResponse): boolean {
  setCorsHeaders(res);
  res.status(204).end();
  return true;
}

File: pages/api/utils/cors.py
"""
CORS (Cross-Origin Resource Sharing) Utility for Python API Routes

Provides centralized CORS configuration for Python-based API endpoints.
Supports configurable allowed origins for production deployments.

Environment Variables:
- ALLOWED_ORIGINS: Comma-separated list of allowed origins
- NEXT_PUBLIC_BASE_URL: Base URL for the application (used as fallback)

Usage:
    from utils.cors import set_cors_headers
    
    def do_GET(self):
        set_cors_headers(self)
        # ... rest of handler
"""

import os


def get_allowed_origin():
    """
    Get allowed origin from environment variables.
    Falls back to wildcard if not configured.
    
    Returns:
        str: Allowed origin value for Access-Control-Allow-Origin header
    """
    # Check for explicit ALLOWED_ORIGINS configuration
    allowed_origins_env = os.environ.get('ALLOWED_ORIGINS', '')
    if allowed_origins_env:
        origins = [origin.strip() for origin in allowed_origins_env.split(',')]
        # For simplicity, use first origin if multiple are specified
        # In production, you'd check the request Origin header
        return origins[0] if origins else '*'
    
    # Check for NEXT_PUBLIC_BASE_URL as fallback
    base_url = os.environ.get('NEXT_PUBLIC_BASE_URL', '')
    if base_url:
        return base_url
    
    # Default to wildcard for development/demo
    return '*'


def set_cors_headers(handler):
    """
    Set CORS headers on a Python API handler response.
    
    Args:
        handler: BaseHTTPRequestHandler instance
    """
    allowed_origin = get_allowed_origin()
    
    handler.send_header('Access-Control-Allow-Origin', allowed_origin)
    handler.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS')
    handler.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization')
    handler.send_header('Access-Control-Max-Age', '86400')
